File: reference/fud-ai/ios/calorietracker/Models/BodyMeasurement.swift
import Foundation

/// One set of tape-measure circumferences logged at a point in time. Every site is optional —
/// the user logs only what they want, and an entry with even a single value is valid. Stored
/// internally in centimetres (display converts to inches when the app is in imperial mode),
/// mirroring how WeightEntry stores kg regardless of the display unit.
///
/// The derived metrics (waist-to-hip, waist-to-height, US-Navy body-fat %, wrist frame size) are
/// computed on the fly from the latest entry plus the profile's height + gender. Nothing here is
/// written back to UserProfile — these are purely extra signal for the AI goal calc and the Coach.
struct BodyMeasurement: Identifiable, Codable, Equatable {
    let id: UUID
    let date: Date
    var neckCm: Double?
    var waistCm: Double?
    var hipsCm: Double?
    var chestCm: Double?
    var upperArmCm: Double?
    var thighCm: Double?
    var calfCm: Double?
    var wristCm: Double?

    init(
        id: UUID = UUID(),
        date: Date = .now,
        neckCm: Double? = nil,
        waistCm: Double? = nil,
        hipsCm: Double? = nil,
        chestCm: Double? = nil,
        upperArmCm: Double? = nil,
        thighCm: Double? = nil,
        calfCm: Double? = nil,
        wristCm: Double? = nil
    ) {
        self.id = id
        self.date = date
        self.neckCm = neckCm
        self.waistCm = waistCm
        self.hipsCm = hipsCm
        self.chestCm = chestCm
        self.upperArmCm = upperArmCm
        self.thighCm = thighCm
        self.calfCm = calfCm
        self.wristCm = wristCm
    }

    /// True when at least one circumference is present — an entry with nothing logged is meaningless.
    var hasAnyValue: Bool {
        [neckCm, waistCm, hipsCm, chestCm, upperArmCm, thighCm, calfCm, wristCm].contains { $0 != nil }
    }

    // MARK: - Derived metrics (all optional — nil when the inputs they need aren't logged)

    /// Waist ÷ hips. WHO cardiometabolic-risk marker. Needs waist + hips.
    var waistToHipRatio: Double? {
        guard let waist = waistCm, let hips = hipsCm, hips > 0 else { return nil }
        return waist / hips
    }

    /// Waist ÷ height. "Keep your waist under half your height." Needs waist + a height.
    func waistToHeightRatio(heightCm: Double) -> Double? {
        guard let waist = waistCm, heightCm > 0 else { return nil }
        return waist / heightCm
    }

    /// U.S. Navy body-fat % estimate (metric coefficients, inputs in cm). Men use neck + waist;
    /// women use neck + waist + hips. Returns nil when the required sites are missing or the
    /// logarithm domain is invalid, and clamps obviously-bad outputs out.
    func usNavyBodyFatPercent(gender: Gender, heightCm: Double) -> Double? {
        guard heightCm > 0, let neck = neckCm, let waist = waistCm else { return nil }
        let result: Double
        switch gender {
        case .female:
            guard let hips = hipsCm else { return nil }
            let inner = waist + hips - neck
            guard inner > 0 else { return nil }
            result = 495.0 / (1.29579 - 0.35004 * log10(inner) + 0.22100 * log10(heightCm)) - 450.0
        case .male, .other:
            let inner = waist - neck
            guard inner > 0 else { return nil }
            result = 495.0 / (1.0324 - 0.19077 * log10(inner) + 0.15456 * log10(heightCm)) - 450.0
        }
        guard result.isFinite, result >= 2, result <= 65 else { return nil }
        return result
    }

    /// Bone-frame size from height ÷ wrist circumference (gender-specific cut-offs). Needs wrist.
    func wristFrame(gender: Gender, heightCm: Double) -> FrameSize? {
        guard let wrist = wristCm, wrist > 0, heightCm > 0 else { return nil }
        let ratio = heightCm / wrist
        switch gender {
        case .female:
            if ratio > 11.0 { return .small }
            if ratio >= 10.1 { return .medium }
            return .large
        case .male, .other:
            if ratio > 10.4 { return .small }
            if ratio >= 9.6 { return .medium }
            return .large
        }
    }

    enum FrameSize: String {
        case small, medium, large
        var label: String {
            switch self {
            case .small: return "Small"
            case .medium: return "Medium"
            case .large: return "Large"
            }
        }
    }

    /// The eight circumference sites, in display order. Used to render the per-site editor rows.
    enum Site: String, CaseIterable, Identifiable {
        case neck, waist, hips, chest, upperArm, thigh, calf, wrist
        var id: String { rawValue }
        var label: String {
            switch self {
            case .neck: return "Neck"
            case .waist: return "Waist"
            case .hips: return "Hips"
            case .chest: return "Chest"
            case .upperArm: return "Upper Arm"
            case .thigh: return "Thigh"
            case .calf: return "Calf"
            case .wrist: return "Wrist"
            }
        }
    }

    func value(for site: Site) -> Double? {
        switch site {
        case .neck: return neckCm
        case .waist: return waistCm
        case .hips: return hipsCm
        case .chest: return chestCm
        case .upperArm: return upperArmCm
        case .thigh: return thighCm
        case .calf: return calfCm
        case .wrist: return wristCm
        }
    }

    /// A copy with one site changed (same id + date — used for in-place daily updates).
    func setting(_ site: Site, to cm: Double?) -> BodyMeasurement {
        var copy = self
        switch site {
        case .neck: copy.neckCm = cm
        case .waist: copy.waistCm = cm
        case .hips: copy.hipsCm = cm
        case .chest: copy.chestCm = cm
        case .upperArm: copy.upperArmCm = cm
        case .thigh: copy.thighCm = cm
        case .calf: copy.calfCm = cm
        case .wrist: copy.wristCm = cm
        }
        return copy
    }

    /// Compact AI-prompt summary of the logged sites + derived metrics, always in cm for a single
    /// consistent unit. Returns nil when nothing is logged so callers can omit the section entirely.
    func promptSummary(gender: Gender, heightCm: Double) -> String? {
        guard hasAnyValue else { return nil }
        var sites: [String] = []
        func site(_ label: String, _ value: Double?) {
            if let value { sites.append("\(label) \(String(format: "%.1f", value)) cm") }
        }
        site("neck", neckCm)
        site("waist", waistCm)
        site("hips", hipsCm)
        site("chest", chestCm)
        site("upper arm", upperArmCm)
        site("thigh", thighCm)
        site("calf", calfCm)
        site("wrist", wristCm)

        var metrics: [String] = []
        if let whr = waistToHipRatio {
            metrics.append("waist-to-hip \(String(format: "%.2f", whr))")
        }
        if let whtr = waistToHeightRatio(heightCm: heightCm) {
            metrics.append("waist-to-height \(String(format: "%.2f", whtr))")
        }
        if let bf = usNavyBodyFatPercent(gender: gender, heightCm: heightCm) {
            metrics.append("US-Navy body fat ~\(String(format: "%.0f", bf))%")
        }
        if let frame = wristFrame(gender: gender, heightCm: heightCm) {
            metrics.append("frame \(frame.label.lowercased())")
        }

        var summary = sites.joined(separator: ", ")
        if !metrics.isEmpty {
            summary += " | derived: " + metrics.joined(separator: ", ")
        }
        return summary
    }
}

File: fud-ai/android/app/src/main/java/com/apoorvdarshan/calorietracker/models/BodyMeasurement.kt
package com.apoorvdarshan.calorietracker.models

import com.apoorvdarshan.calorietracker.R
import kotlinx.serialization.Serializable
import java.time.Instant
import java.util.Locale
import java.util.UUID
import kotlin.math.log10

/**
 * One set of tape-measure circumferences logged at a point in time. Every site is optional — the
 * user logs only what they want. Stored internally in centimetres (display converts to inches when
 * the app is in imperial mode), mirroring how WeightEntry stores kg regardless of display unit.
 *
 * The derived metrics (waist-to-hip, waist-to-height, US-Navy body-fat %, wrist frame size) are
 * computed on the fly from this entry plus the profile's height + gender. Nothing here is written
 * back to UserProfile — these are purely extra signal for the AI goal calc and the Coach.
 */
@Serializable
data class BodyMeasurement(
    @Serializable(with = UuidSerializer::class)
    val id: UUID = UUID.randomUUID(),
    @Serializable(with = InstantSerializer::class)
    val date: Instant = Instant.now(),
    val neckCm: Double? = null,
    val waistCm: Double? = null,
    val hipsCm: Double? = null,
    val chestCm: Double? = null,
    val upperArmCm: Double? = null,
    val thighCm: Double? = null,
    val calfCm: Double? = null,
    val wristCm: Double? = null
) {
    /** True when at least one circumference is present — an empty entry is meaningless. */
    val hasAnyValue: Boolean
        get() = listOf(neckCm, waistCm, hipsCm, chestCm, upperArmCm, thighCm, calfCm, wristCm)
            .any { it != null }

    /** Waist ÷ hips. WHO cardiometabolic-risk marker. Needs waist + hips. */
    val waistToHipRatio: Double?
        get() {
            val waist = waistCm ?: return null
            val hips = hipsCm ?: return null
            if (hips <= 0) return null
            return waist / hips
        }

    /** Waist ÷ height. "Keep your waist under half your height." Needs waist + a height. */
    fun waistToHeightRatio(heightCm: Double): Double? {
        val waist = waistCm ?: return null
        if (heightCm <= 0) return null
        return waist / heightCm
    }

    /**
     * U.S. Navy body-fat % estimate (metric coefficients, inputs in cm). Men use neck + waist;
     * women use neck + waist + hips. Returns null when required sites are missing or the logarithm
     * domain is invalid, and rejects obviously-bad outputs.
     */
    fun usNavyBodyFatPercent(gender: Gender, heightCm: Double): Double? {
        if (heightCm <= 0) return null
        val neck = neckCm ?: return null
        val waist = waistCm ?: return null
        val result: Double = when (gender) {
            Gender.FEMALE -> {
                val hips = hipsCm ?: return null
                val inner = waist + hips - neck
                if (inner <= 0) return null
                495.0 / (1.29579 - 0.35004 * log10(inner) + 0.22100 * log10(heightCm)) - 450.0
            }
            else -> {
                val inner = waist - neck
                if (inner <= 0) return null
                495.0 / (1.0324 - 0.19077 * log10(inner) + 0.15456 * log10(heightCm)) - 450.0
            }
        }
        if (!result.isFinite() || result < 2 || result > 65) return null
        return result
    }

    /** Bone-frame size from height ÷ wrist circumference (gender-specific cut-offs). Needs wrist. */
    fun wristFrame(gender: Gender, heightCm: Double): FrameSize? {
        val wrist = wristCm ?: return null
        if (wrist <= 0 || heightCm <= 0) return null
        val ratio = heightCm / wrist
        return when (gender) {
            Gender.FEMALE -> when {
                ratio > 11.0 -> FrameSize.SMALL
                ratio >= 10.1 -> FrameSize.MEDIUM
                else -> FrameSize.LARGE
            }
            else -> when {
                ratio > 10.4 -> FrameSize.SMALL
                ratio >= 9.6 -> FrameSize.MEDIUM
                else -> FrameSize.LARGE
            }
        }
    }

    /**
     * Compact AI-prompt summary of the logged sites + derived metrics, always in cm for a single
     * consistent unit. Returns null when nothing is logged so callers can omit the section entirely.
     */
    fun promptSummary(gender: Gender, heightCm: Double): String? {
        if (!hasAnyValue) return null
        val sites = mutableListOf<String>()
        fun site(label: String, value: Double?) {
            if (value != null) sites.add("$label ${String.format(Locale.US, "%.1f", value)} cm")
        }
        site("neck", neckCm)
        site("waist", waistCm)
        site("hips", hipsCm)
        site("chest", chestCm)
        site("upper arm", upperArmCm)
        site("thigh", thighCm)
        site("calf", calfCm)
        site("wrist", wristCm)

        val metrics = mutableListOf<String>()
        waistToHipRatio?.let { metrics.add("waist-to-hip ${String.format(Locale.US, "%.2f", it)}") }
        waistToHeightRatio(heightCm)?.let { metrics.add("waist-to-height ${String.format(Locale.US, "%.2f", it)}") }
        usNavyBodyFatPercent(gender, heightCm)?.let { metrics.add("US-Navy body fat ~${String.format(Locale.US, "%.0f", it)}%") }
        wristFrame(gender, heightCm)?.let { metrics.add("frame ${it.label.lowercase()}") }

        var summary = sites.joinToString(", ")
        if (metrics.isNotEmpty()) summary += " | derived: " + metrics.joinToString(", ")
        return summary
    }

    enum class FrameSize(val label: String, val labelRes: Int) {
        SMALL("Small", R.string.frame_small),
        MEDIUM("Medium", R.string.frame_medium),
        LARGE("Large", R.string.frame_large)
    }

    /** The eight circumference sites, in display order. Used to render the per-site editor rows. */
    enum class Site(val label: String, val labelRes: Int) {
        NECK("Neck", R.string.measure_neck), WAIST("Waist", R.string.measure_waist),
        HIPS("Hips", R.string.measure_hips), CHEST("Chest", R.string.measure_chest),
        UPPER_ARM("Upper Arm", R.string.measure_upper_arm), THIGH("Thigh", R.string.measure_thigh),
        CALF("Calf", R.string.measure_calf), WRIST("Wrist", R.string.measure_wrist)
    }

    fun value(site: Site): Double? = when (site) {
        Site.NECK -> neckCm
        Site.WAIST -> waistCm
        Site.HIPS -> hipsCm
        Site.CHEST -> chestCm
        Site.UPPER_ARM -> upperArmCm
        Site.THIGH -> thighCm
        Site.CALF -> calfCm
        Site.WRIST -> wristCm
    }

    /** A copy with one site changed (same id + date — used for in-place daily updates). */
    fun setting(site: Site, cm: Double?): BodyMeasurement = when (site) {
        Site.NECK -> copy(neckCm = cm)
        Site.WAIST -> copy(waistCm = cm)
        Site.HIPS -> copy(hipsCm = cm)
        Site.CHEST -> copy(chestCm = cm)
        Site.UPPER_ARM -> copy(upperArmCm = cm)
        Site.THIGH -> copy(thighCm = cm)
        Site.CALF -> copy(calfCm = cm)
        Site.WRIST -> copy(wristCm = cm)
    }
}
